### PR TITLE
[Trilinos] Build More Products

### PR DESCRIPTION
**Pull request by Wimmerer**
Currently Trilinos is just used as a dependency for xyce_jll as far as I can tell.

I'd like to start generating some wrappers for some of these libraries, though. As part of that I wanted the greatest subset of libraries built without pulling in more dependencies.

This PR *attempts to* maintain the previous capability while building all other core packages except:
1. SEACAS
2. MiniTensor
3. Claps
4. Trios
5. TriKota (Dakota Toolkit)
6. (parts of) ShyLU
7. STK (Sierra Toolkit Mesh)
8. (parts of?) Panzer
9. (parts of?) Pike
10. Zoltan2 (optional dependency of several built packages)

Most of the above are missing due to MPI, Boost, or DLlib. I cannot build Zoltan2 due to some issue with libamd I believe, needs further investigation. 

In terms of third party libraries it looks like I managed to use BLAS and LAPACK, as well as AMD, Cholmod and UMFPACK. There are *many many* more third party libraries we may wish to build (MKL, Pardiso, CUDA, METIS, Pthreads, PETsc, TBB, Eigen, etc).

Some notes:
1. Nothing is MPI enabled. A future package should allow users to supply an MPI enabled Trilinos.
2. Tpetra is run in Serial mode. The other execution spaces are Threads (Pthreads I think), OpenMP, and CUDA. I believe this is because we don't find Pthreads or OpenMP. This can be rectified, but I don't know if it interferes with threading from Julia/BLAS. Not sure how that effects Epetra and other packages.
3. I probably missed removing some non-Trilinos LibraryProducts.
4. Zoltan2 missing might cause some issues, however it is listed as optional for all the dependents.

The big thing is really to ensure that xyce_jll doesn't break with the changes here. I don't believe it should, but I don't have any easy method to test that.

### CODE REVIEW

**Comment by Wimmerer** [comment]
Hmm. Failing in different places on different platforms is very fun indeed. Compiles on x86...

Why does [this](https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=12560&view=logs&j=e07d7dc6-9270-57d1-72f1-2d2a416920fb&t=6d48680b-fe95-587c-373a-9a4c04a1e5b9&l=4185) happen on freebsd but not on linux?

**Comment by giordano** [comment]
> Why does this happen on freebsd but not on linux?

Please copy here the error messages.  Azure logs are ephemeral and not searchable from here.  To answer your question: I haven't really dug into the specific error, but in general keep in mind that different compilers have different opinions on what constitutes an error.

**Comment by Wimmerer** [comment]
Here's the very beautiful error for i686:

```
int atomic_fetch_sub( volatile int * const dest , const int val )
     ^~~~~~~~~~~~~~~~
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Fetch_Sub.hpp:138:5: note:   no known conversion for argument 1 from ‘volatile long long int*’ to ‘volatile int*’
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Fetch_Sub.hpp:142:10: note: candidate: long int Kokkos::atomic_fetch_sub(volatile long int*, long int)
 long int atomic_fetch_sub( volatile long int * const dest , const long int val )
          ^~~~~~~~~~~~~~~~
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Fetch_Sub.hpp:142:10: note:   no known conversion for argument 1 from ‘volatile long long int*’ to ‘volatile long int*’
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Fetch_Sub.hpp:148:14: note: candidate: unsigned int Kokkos::atomic_fetch_sub(volatile unsigned int*, unsigned int)
 unsigned int atomic_fetch_sub( volatile unsigned int * const dest , const unsigned int val )
              ^~~~~~~~~~~~~~~~
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Fetch_Sub.hpp:148:14: note:   no known conversion for argument 1 from ‘volatile long long int*’ to ‘volatile unsigned int*’
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Fetch_Sub.hpp:152:19: note: candidate: long unsigned int Kokkos::atomic_fetch_sub(volatile long unsigned int*, long unsigned int)
 unsigned long int atomic_fetch_sub( volatile unsigned long int * const dest , const unsigned long int val )
                   ^~~~~~~~~~~~~~~~
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Fetch_Sub.hpp:152:19: note:   no known conversion for argument 1 from ‘volatile long long int*’ to ‘volatile long unsigned int*’
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Fetch_Sub.hpp:159:3: note: candidate: template<class T> T Kokkos::atomic_fetch_sub(volatile T*, typename Kokkos::Impl::enable_if<(sizeof (T) == sizeof (int)), const T>::type)
 T atomic_fetch_sub( volatile T * const dest ,
   ^~~~~~~~~~~~~~~~
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Fetch_Sub.hpp:159:3: note:   template argument deduction/substitution failed:
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Fetch_Sub.hpp: In substitution of ‘template<class T> T Kokkos::atomic_fetch_sub(volatile T*, typename Kokkos::Impl::enable_if<(sizeof (T) == sizeof (int)), const T>::type) [with T = long long int]’:
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Decrement.hpp:109:32:   required from here
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Fetch_Sub.hpp:159:3: error: invalid use of incomplete type ‘struct Kokkos::Impl::enable_if<false, const long long int>’
In file included from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_MemoryTraits.hpp:47:0,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_HostSpace.hpp:54,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_View.hpp:53,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_Parallel.hpp:52,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_Serial.hpp:55,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_Core.hpp:53,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Core.cpp:44:
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Traits.hpp:209:8: note: declaration of ‘struct Kokkos::Impl::enable_if<false, const long long int>’
 struct enable_if ;
        ^~~~~~~~~
In file included from /workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Decrement.hpp:48:0,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_Atomic.hpp:242,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_View.hpp:47,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_ViewMapping.hpp:56,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_View.hpp:372,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_Parallel.hpp:52,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_Serial.hpp:55,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_Core.hpp:53,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Core.cpp:44:
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Fetch_Sub.hpp:177:3: note: candidate: template<class T> T Kokkos::atomic_fetch_sub(volatile T*, typename Kokkos::Impl::enable_if<((sizeof (T) != sizeof (int)) && (sizeof (T) == sizeof (long int))), const T>::type)
 T atomic_fetch_sub( volatile T * const dest ,
   ^~~~~~~~~~~~~~~~
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Fetch_Sub.hpp:177:3: note:   template argument deduction/substitution failed:
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Fetch_Sub.hpp: In substitution of ‘template<class T> T Kokkos::atomic_fetch_sub(volatile T*, typename Kokkos::Impl::enable_if<((sizeof (T) != sizeof (int)) && (sizeof (T) == sizeof (long int))), const T>::type) [with T = long long int]’:
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Decrement.hpp:109:32:   required from here
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Fetch_Sub.hpp:177:3: error: invalid use of incomplete type ‘struct Kokkos::Impl::enable_if<false, const long long int>’
In file included from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_MemoryTraits.hpp:47:0,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_HostSpace.hpp:54,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_View.hpp:53,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_Parallel.hpp:52,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_Serial.hpp:55,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_Core.hpp:53,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Core.cpp:44:
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Traits.hpp:209:8: note: declaration of ‘struct Kokkos::Impl::enable_if<false, const long long int>’
 struct enable_if ;
        ^~~~~~~~~
In file included from /workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Decrement.hpp:48:0,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_Atomic.hpp:242,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_View.hpp:47,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_ViewMapping.hpp:56,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_View.hpp:372,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_Parallel.hpp:52,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_Serial.hpp:55,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_Core.hpp:53,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Core.cpp:44:
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Fetch_Sub.hpp:199:3: note: candidate: template<class T> T Kokkos::atomic_fetch_sub(volatile T*, typename Kokkos::Impl::enable_if<((sizeof (T) != 4) && (sizeof (T) != 8)), const T>::type&)
 T atomic_fetch_sub( volatile T * const dest ,
   ^~~~~~~~~~~~~~~~
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Fetch_Sub.hpp:199:3: note:   template argument deduction/substitution failed:
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Fetch_Sub.hpp: In substitution of ‘template<class T> T Kokkos::atomic_fetch_sub(volatile T*, typename Kokkos::Impl::enable_if<((sizeof (T) != 4) && (sizeof (T) != 8)), const T>::type&) [with T = long long int]’:
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Decrement.hpp:109:32:   required from here
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Atomic_Fetch_Sub.hpp:199:3: error: invalid use of incomplete type ‘struct Kokkos::Impl::enable_if<false, const long long int>’
In file included from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_MemoryTraits.hpp:47:0,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_HostSpace.hpp:54,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_View.hpp:53,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_Parallel.hpp:52,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_Serial.hpp:55,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_Core.hpp:53,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Core.cpp:44:
/workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Traits.hpp:209:8: note: declaration of ‘struct Kokkos::Impl::enable_if<false, const long long int>’
 struct enable_if ;
        ^~~~~~~~~
In file included from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_Parallel_Reduce.hpp:47:0,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_Parallel.hpp:253,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_Serial.hpp:55,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_Core.hpp:53,
                 from /workspace/srcdir/Trilinos/packages/kokkos/core/src/impl/Kokkos_Core.cpp:44:
/workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_NumericTraits.hpp: In static member function ‘static constexpr long int Kokkos::reduction_identity<long int>::max()’:
/workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_NumericTraits.hpp:111:68: warning: overflow in implicit constant conversion [-Woverflow]
   KOKKOS_FORCEINLINE_FUNCTION constexpr static long max()  {return LLONG_MIN;}
                                                                    ^~~~~~~~~
/workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_NumericTraits.hpp: In static member function ‘static constexpr long int Kokkos::reduction_identity<long int>::min()’:
/workspace/srcdir/Trilinos/packages/kokkos/core/src/Kokkos_NumericTraits.hpp:112:68: warning: overflow in implicit constant conversion [-Woverflow]
   KOKKOS_FORCEINLINE_FUNCTION constexpr static long min()  {return LLONG_MAX;}
                                                                    ^~~~~~~~~
make[2]: *** [packages/kokkos/core/src/CMakeFiles/kokkoscore.dir/build.make:96: packages/kokkos/core/src/CMakeFiles/kokkoscore.dir/impl/Kokkos_Core.cpp.o] Error 1
```

**Comment by giordano** [comment]
Considering it fails on 32-bit platforms I have a strong suspect on an upstream bug

File: T/Trilinos/build_tarballs.jl [hunk 1]
```diff
@@ -10,16 +10,49 @@ sources = [
...
+-DAMD_LIBRARY_NAMES="libsuitesparseconfig.${dlext};libamd.${dlext};libklu.${dlext};libcolamd.${dlext};libbtf.${dlext}" \
+-DTPL_AMD_INCLUDE_DIRS="${prefix}/include" \
+-DTPL_ENABLE_UMFPACK=ON \
+-DUMFPACK_LIBRARY_DIRS="${prefix/lib}" \
+-DUMFPACK_LIBRARY_NAMES="libsuitesparseconfig.${dlext};libamd.${dlext};libumfpack.${dlext}" \
+-DTPL_UMFPACK_INCLUDE_DIRS="${prefix}/include" \
+-DTPL_ENABLE_Cholmod=ON \
+-DCholmod_LIBRARY_DIRS="${prefix/lib}" \
+-DCholmod_LIBRARY_NAMES="libsuitesparseconfig.${dlext};libamd.${dlext};libcholmod.${dlext};libcolamd.${dlext}" \
+-DBLAS_LIBRARY_DIRS="${prefix}/lib" \
```
Comment (giordano) [inline T/Trilinos/build_tarballs.jl:42]:
```suggestion
-DBLAS_LIBRARY_DIRS="${libdir}" \
```

File: T/Trilinos/build_tarballs.jl [hunk 2]
```diff
@@ -10,16 +10,49 @@ sources = [
...
+-DBLAS_LIBRARY_NAMES="libopenblas.${dlext}" \
+-DLAPACK_LIBRARY_DIRS="${prefix}/lib" \
```
Comment (giordano) [inline T/Trilinos/build_tarballs.jl:44]:
```suggestion
-DLAPACK_LIBRARY_DIRS="${libdir}" \
```

File: T/Trilinos/build_tarballs.jl [hunk 3]
```diff
@@ -67,4 +169,4 @@ dependencies = [
 ]
 
 # Build the tarballs, and possibly a `build.jl` as well.
-build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; preferred_gcc_version = v"7.1.0")
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; preferred_gcc_version = v"7.2.0")
```
Comment (giordano) [inline T/Trilinos/build_tarballs.jl:172]: Why this change?  We don't have 7.2, this goes straight to 8

Comment (Wimmerer) [reply T/Trilinos/build_tarballs.jl:169]: I was attempting to match Kokkos' tested compiler set. I can change it back.

File: T/Trilinos/build_tarballs.jl [hunk 4]
```diff
@@ -10,16 +10,49 @@ sources = [
...
-make -j${nprocs}
+cmake \
+-DCMAKE_CXX_FLAGS="$FLAGS" -DCMAKE_C_FLAGS="$FLAGS" \
+-DCMAKE_Fortran_FLAGS="$FLAGS" \
+-DCMAKE_C_COMPILER=gcc \
+-DCMAKE_CXX_COMPILER=g++ \
+-DCMAKE_Fortran_COMPILER=gfortran \
```
[the remark this answers is not in the document]
Comment (giordano) [reply T/Trilinos/build_tarballs.jl]: This still _must_ use the toolchain file.  Also, it'd be nice to have the flags indented by 4 spaces, at the moment this is rather unreadable

**Comment by Wimmerer** [comment]
I strongly suspect 32-bit isn't officially supported (based on some issues upstream) and that the previous build which worked with 32-bit was a fluke (subset of packages which did support 32-bit).

File: T/Trilinos/build_tarballs.jl [hunk 5]
```diff
@@ -10,16 +10,47 @@ sources = [
...
-make -j${nprocs}
+cmake \
+    -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN}
```
Comment (giordano) [inline T/Trilinos/build_tarballs.jl:21]:
```suggestion
    -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} \
```

**Comment by Wimmerer** [comment]
We're back to the manually specified compilers. I've enabled OpenMP. If this works on 64 bit platforms I'll split it out into two builders with one `common.jl` that builds the Epetra stack (the original stack from the first builder), and then one that builds the Tpetra stack.

I'll probably ensure the platforms are disjoint rather than built Tpetra stack based on the Epetra one.

**Comment by Wimmerer** [comment]
Since people tangentially interested in Trilinos think 32-bit support is at best nice to have, I'd like to see if we can just disable 32-bit. A 32-bit build will have to be restricted _only_ to 32-bit, since many packages have optional dependencies on the new 64-bit stuff. Or put another way `Trilinos_Tpetra_jll` won't be able to depend on `Trilinos_Epetra_jll`, _and_ will conflict.

If a 32-bit build becomes necessary we can add it, but it will be separate from this one, since only like 5 lines would be shared between them. Thoughts @giordano?

**Comment by Wimmerer** [comment]
Any way to filter 32 bit platforms?

**Comment by giordano** [comment]
> Any way to filter 32 bit platforms?

```julia
nbits(p) != 32
```

**Comment by Wimmerer** [comment]
@giordano I don't think we should do the common/multi builder thing here. Maybe if conditional deps become available, but without them I don't think this will work. 

A 32-bit build would not work as a dependency of the 64-bit build, since the 64-bit one would have to rebuild a significant fraction of the 32-bit libraries. I think it's reasonable to have this one build just for 64-bit, if we get conditional dependencies in the future we can add a 32-bit only builder, and Trilinos.jl can have both be conditional deps.

**Comment by Wimmerer** [comment]
https://github.com/kokkos/kokkos/issues/2873 :(

**Comment by Wimmerer** [comment]
@giordano, musl is not officially supported either. 

Perhaps it's better than I make Trilinos.jl and require users install it themselves.
